### DISCUSSION TITLE
xrdp + RealVNC: No configured security type is supported by 3.3 VNC Viewer

### DISCUSSION THREAD

**Discussion by usa-** [neutrinolabs/xrdp#3203] [Q&A]
I am connecting to Raspberry Pi 4, Raspberry Pi OS (Debian based) Bookworm 64-bit.
I have switched the backend to X11 from Wayland, because RealVNC does not support the latter.

Using Remote Desktop client, I can connect to xrdp with configuration [Xorg]. An empty virtual desktop shows up.
Using any VNC client, I can connect to RealVNC. A physical (active) desktop shows up with previously opened windows, which I also can see on the RPi screen.
But using Remote Desktop client, I can not connect to xrdp with configuration [Xvnc] to see the physical (active) desktop, like I did some years ago.

I changed the port to 5900, so that configuration looks like this:
```
[Xvnc]
name=Xvnc
lib=libvnc.so
username=ask
password=ask
ip=127.0.0.1
port=5900
```

No matter which Encryption (Always on, Prefer on, Prefer off) and Authentication (VNC password, Unix password, None) I set, I get the following in the log files (timestamps stripped).

xrdp
```
[INFO ] loaded module 'libvnc.so' ok, interface size 4064, version 4
[INFO ] VNC started connecting 
[INFO ] VNC connecting to 127.0.0.1 5900 
[INFO ] VNC tcp connected 
[INFO ] VNC security level is 0 (1 = none, 2 = standard) 
[ERROR] VNC Server will disconnect 
[ERROR] VNC error 1 after security negotiation 
[ERROR] VNC error before sending share flag 
[ERROR] VNC error before receiving server init 
[ERROR] VNC error before receiving pixel format 
[ERROR] VNC error before receiving name length 
[ERROR] VNC error before receiving name 
[INFO ] VNC error - problem connecting 
[INFO ] some problem 
[ERROR] xrdp_wm_log_msg: Error connecting to user session 
[INFO ] Error connecting to user session
```

RealVNC
```
IpFilter: ACCEPT [::1] 
Connections: connected: ::1::46846 (TCP) 
SConnection: reading protocol version 
SConnection: Client needs protocol version 3.3 
RfbConnection: Peer does not support sync, KeepAlive disabled 
Connections: disconnected: ::1::46846 (TCP) ([ConnFailed] No configured security type is supported by 3.3 VNC Viewer)
FdIo: shutdown 0x55b7761c70 (fd=17), flush true (tellHandler true) 
FdIo: shutdown complete 0x55b7761c70 (fd=17) 
FdIo: shutdown 0x55b7761c70 (fd=17), flush true (tellHandler false)
```

**Comment by matt335672** [comment]
The security level is sent from the RealVNC server to xrdp, not the other way round.

A value of zero is sent by the server to xrdp if it doesn't support either 'none' or 'VNC authentication'.

My guess is you have to configure the RealVNC server on the PI to allow for VNC authentication. Have a look at this page:-

https://raspberrypi.stackexchange.com/questions/69136/vnc-error-no-configured-security-type-is-supported-by-3-3-vnc-viewer

**Reply by usa-** [reply]
Thank you, I have read this thread before posting.
My post states that I have tried the methods suggested there.

I think I have found the problem. If I leave other RealVNC server settings intact, no matter which authentication type I set in the server, a client with protocol version 3.3 fails to connect with the same message: "No configured security type is supported by 3.3 VNC Viewer". Even if I connect with RealVNC viewer and limit the protocol version to 3.3 in it.

P. S. I ended up with uninstalling RealVNC and installing x11vnc.
Now, depending on x11vnc command-line switches, I can connect **with VNC client** either without password, or with the VNC password.
Or **with Remote Desktop Client via XRDP** either with any username and any password, or with any username and the VNC password.

Is there a way to make XRDP perform the authentication with Unix username and password, and then connect to VNC at localhost with no password?

**Reply by matt335672** [reply]
Yes there is.

Rather than using `username=` and `password=`, use `pamusername=` and `pampassword=` for authentication and set the password to be the VNC password. I can't remember if that's allowed to be empty or not.

**Reply by usa-** [reply]
Sorry, I don't understand your idea completely.
Can you provide a full [Xvnc] section after the changes?
You can see my current [Xvnc] section in the original post.
1. Assume my x11vnc password is "hello", and the connecting user can authenticate as any OS user.
2. Assume my x11vnc has no password, and the connecting user must authenticate as the OS user "pi".

**Reply by matt335672** [reply]
Off the top of my head:-
```
[Xvnc]
name=Xvnc
lib=libvnc.so
pamusername=ask
pampassword=ask
password=hello
ip=127.0.0.1
port=5900
```

```
[Xvnc]
name=Xvnc
lib=libvnc.so
pamusername=pi
pampassword=ask
password=
ip=127.0.0.1
port=5900
```

I'm not sure an empty password will be accepted in option 2.

**Reply by usa-** [reply]
Option 1.
VNC server is set to run with the password "hello". VNC clients can connect normally with the password "hello". XRDP can not:
```
[INFO ] connecting to PAM authenticator on 127.0.0.1:3350
[INFO ] xrdp_wm_log_msg: PAM authenticator connect ok
[INFO ] PAM authenticator connect ok
[INFO ] xrdp_wm_log_msg: Performing access control for ask
[INFO ] Performing access control for ask
[INFO ] sending login info to session manager, please wait...
[INFO ] xrdp_wm_log_msg: Reply from access control: Authentication failure
[INFO ] Reply from access control: Authentication failure
```
Adding `user=user` and/or changing to `pamusername=pi` does not help.

Option 2.
VNC server is set to run with no password. VNC client can connect normally without a password. XRDP can not:
```
[INFO ] connecting to PAM authenticator on 127.0.0.1:3350
[INFO ] xrdp_wm_log_msg: PAM authenticator connect ok
[INFO ] PAM authenticator connect ok
[INFO ] xrdp_wm_log_msg: Performing access control for pi
[INFO ] Performing access control for pi
[INFO ] sending login info to session manager, please wait...
[INFO ] xrdp_wm_log_msg: Reply from access control: Authentication failure
[INFO ] Reply from access control: Authentication failure
```
Changing to `password=password`, removing this variable or adding `user=user` does not help.

**Reply by matt335672** [reply]
This is on my test VM (I couldn't look earlier). I'm running a development version of xrdp, but it should also be fine for v0.10.x. It also works on the latest v0.9 wavefront.

```
[localvnc]
name=localvnc
lib=libvnc.so
ip=localhost
port=5900
#username=ask
password=
pamusername=ask
pampassword=ask
# Needed for xrdp v0.9.x
#pamsessionmng=127.0.0.1
# This line needs xrdp 0.9.16 or later
chansrvport=DISPLAY(0)
```

**Reply by usa-** [reply]
So basically your configuration is the same as mine, except `chansrvport=DISPLAY(0)` which must not affect connection to Xvnc.
I use xrdp version 0.9.21.1-1 arm64, which is included in Raspberry Pi OS repositiory.
Now I will compile the latest version from here and tell you the result.

**Reply by usa-** [reply]
I installed xrdp 0.10.80 using the instructions provided [here](https://github.com/neutrinolabs/xrdp/wiki/Building-on-Debian-8).
I did not adapt the configure line.
I did not compile and install xorgxrdp.

After installation, I removed all the default connection configurations from xrdp.ini and added the one provided in your last comment.
RDP connection fails with the following messages in the log file:
```
[INFO ] Client supports single-screen resizes by xrdp
[INFO ] xrdp_caps_process_pointer: client supports new(color) cursor
[INFO ] Client supports Unicode input
[INFO ] xrdp_process_offscreen_bmpcache: support level 1 cache size 7864320 MB cache entries 512
[WARN ] Client Capability: not enough orders supported by client, client wants off screen bitmap but offscreen bitmaps disabled
[INFO ] Client Capability: LARGE_POINTER_FLAG_96x96 supported
[INFO ] Loading keymap file /etc/xrdp/km-00000409.toml
[INFO ] Login screen monitor height is 900 pixels over 179 mm (127 DPI)
[WARN ] Ignoring obsolete SCP port value '3350'
[INFO ] xrdp_wm_log_msg: sesman connect ok
[INFO ] xrdp_wm_log_msg: Performing access control for ask
[INFO ] xrdp_wm_log_msg: User does not exist, or could not be authenticated
```
What am I doing wrong?

P. S. `cache size 7864320 MB` - wow, my Raspberry only has 8000 MB of RAM.

P. P. S. The connection only fails if I use autologon in my RDP client. If I remove it, connect and fill the pamusername and pampassord fields manually, I see a black screen for about 8 seconds, then the desktop appears without the taskbar, and it does not react to my keyboard or mouse. After I disconnect and reconnect, I see that it actually reacted, I just cant see the results in the same session.

**Reply by matt335672** [reply]
I've had a look at the way the code works, and I'm afraid you can't use autologon. If you send a password from the client, xrdp will insist on using that for the VNC connection, and that won't work. 

However, you can save a bit of typing by setting the `pamusername` to `user` or `askuser` and it will either set the username to `user`, or pre-set it and let you change it. Try those.

Other things:-
1) Make sure screen blanking is disabled (see #1701)
2) To get the chansrv functionality, you'll need to start xrdp-chansrv running as part of your session somehow. There are various ways to do this.

Thanks for pointing out the bad message! That value is 7680 kilobytes ( = 7864320 bytes), so it's being mis-reported. I'll fix that in devel.

**Reply by matt335672** [reply]
Reporting fix in neutrinolabs/xrdp#3209
